camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/DslDeferredFunctionCall.java

```diff
@@ -0,0 +1,169 @@
+
+import org.apache.brooklyn.api.mgmt.Task;
+import org.apache.brooklyn.camp.brooklyn.spi.dsl.methods.BrooklynDslCommon;
+import org.apache.brooklyn.core.entity.EntityInternal;
+import org.apache.brooklyn.core.mgmt.BrooklynTaskTags;
+import org.apache.brooklyn.util.core.task.Tasks;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.javalang.Reflections;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Objects;
+
+public class DslDeferredFunctionCall extends BrooklynDslDeferredSupplier<Object> {
+
+    private static final long serialVersionUID = 3243262633795112155L;
+
+    // TODO should this be some of the super types?
+    private BrooklynDslDeferredSupplier<?> object;
+    private String fnName;
+    private List<?> args;
+
+    public DslDeferredFunctionCall(BrooklynDslDeferredSupplier<?> o, String fn, List<Object> args) {
+        this.object = o;
+        this.fnName = fn;
+        this.args = args;
+    }
+
+    @Override
+    public Maybe<Object> getImmediately() {
+        Maybe<?> obj = resolveImmediate(object);
+        if (obj.isPresent()) {
+            if (obj.isNull()) {
+                throw new IllegalArgumentException("Deferred function call, " + object + 
+                        " evaluates to null (when calling " + fnName + "(" + toString(args) + "))");
+            }
+            return Maybe.of(invokeOn(obj.get()));
+        }
+        return Maybe.absent("Could not evaluate immediately " + object);
+    }
+
+    @Override
+    public Task<Object> newTask() {
+        return Tasks.builder()
+                .displayName("Deferred function call " + object + "." + fnName + "(" + toString(args) + ")")
+                .tag(BrooklynTaskTags.TRANSIENT_TASK_TAG)
+                .dynamic(false)
+                .body(new Callable<Object>() {
+                    @Override
+                    public Object call() throws Exception {
+                        Object obj = DslDeferredFunctionCall.this.resolveBlocking(object).orNull();
+                        if (obj == null) {
+                            throw new IllegalArgumentException("Deferred function call, " + object + 
+                                    " evaluates to null (when calling " + fnName + "(" + DslDeferredFunctionCall.toString(args) + "))");
+                        }
+                        return invokeOn(obj);
+                    }
+
+                }).build();
+    }
+
+    protected Maybe<?> resolveImmediate(Object object) {
+        return resolve(object, true);
+    }
+    protected Maybe<?> resolveBlocking(Object object) {
+        return resolve(object, false);
+    }
+    protected Maybe<?> resolve(Object object, boolean immediate) {
+        if (object instanceof DslCallable || object == null) {
+            return Maybe.of(object);
+        }
+        Maybe<?> resultMaybe = Tasks.resolving(object, Object.class)
```

Comment:
How is this code reachable?  This class is only constructed when `object` is `DslCallable` in which case the above branch applies, isn't it?